modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IncrementalSnapshotVerificationTask.java

```diff
@@ -0,0 +1,421 @@
+                            int valHash = dataEntry.key().hashCode() + Arrays.hashCode(dataEntry.value().valueBytes(null));
+                            int verHash = dataEntry.writeVersion().hashCode();
+
+                            if (v == null)
+                                v = new PartitionHashHolder();
+
+                            v.increment(valHash, verHash);
+
+                            return v;
+                        }
+                        catch (IgniteCheckedException ex) {
+                            exceptions.add(ex);
+
+                            return null;
+                        }
+                    });
+                }, txRec -> {
+                    if (txRec.state() == TransactionState.PREPARED) {
+                        // Collect only primary nodes. For some cases backup nodes is included into TxRecord#participationNodes()
+                        // but actually doesn't even start transaction, for example, if the node participates only as a backup
+                        // of reading only keys.
+                        Set<Short> primPartNodes = txRec.participatingNodes().keySet();
+
+                        if (primPartNodes.contains(locShortId)) {
+                            txPrimPartNodes.put(txRec.nearXidVersion(), primPartNodes);
+                            activeDhtTxs.add(txRec.writeVersion());
+                        }
+                        else {
+                            for (Collection<Short> backups: txRec.participatingNodes().values()) {
+                                if (backups.contains(ALL_NODES) || backups.contains(locShortId))
+                                    activeDhtTxs.add(txRec.writeVersion());
+                            }
+                        }
+                    }
+                    else if (txRec.state() == TransactionState.COMMITTED) {
+                        activeDhtTxs.remove(txRec.writeVersion());
+
+                        Set<Short> partNodes = txPrimPartNodes.remove(txRec.nearXidVersion());
+
+                        // Legal cases:
+                        // 1. This node is a transaction near node, but not primary or backup node.
+                        // 2. This node participated in the transaction multiple times (e.g., primary for one key and backup for other key).
+                        // 3. A transaction is included into previous incremental snapshot.
+                        if (partNodes == null)
+                            return;
+
+                        calcTransactionHash.accept(txRec.nearXidVersion(), partNodes);
+                    }
+                    else if (txRec.state() == TransactionState.ROLLED_BACK) {
+                        activeDhtTxs.remove(txRec.writeVersion());
+                        txPrimPartNodes.remove(txRec.nearXidVersion());
+                    }
+                });
+
+                // All active transactions that didn't log COMMITTED or ROLL_BACK records are considered committed.
```

Comment:
Can transactions without WAL record be rolled back during node recovery?
In case node fails after incremental snapshot created but before TX_RECORD loged